# Brand vs Mileage — Exploring eBay's Used Car Sales

## Introduction

In this project, I will be using the dataset for used cars from eBay Kleinanzeigen, a classified ads section of the German eBay website. The dataset was originally scraped and uploaded to Kaggle, but the version we will be using was altered to resemble more of what a scraped dataset would look like.

The goal is to clean the data and analyze the used car listings to see how brands and mileage play a role in their pricing.

The data dictionary provided with data is as follows:


|Column label | What is represents |
|:-|:-|
|dateCrawled|When this ad was first crawled|
|name|Name of the car|
|seller|Whether the seller is private or a dealer|
|offerType|The type of listing|
|price|The price on the ad to sell the car|
|abtest|Whether the listing is included in an A/B test|
|vehicleType|The vehicle type|
|yearOfRegistration|The year in which the car was first registered|
|gearbox|The transmission type|
|powerPS|The power of the car in PS|
|model|The car model name|
|kilometer|How many kilometers the car has driven|
|monthOfRegistration|The month in which the car was first registered|
|fueltype|What type of fuel the car uses|
|brand|The brand of the car|
|notRepairedDamage|If the car has a damage which is not yet repaired|
|dateCreated|The date on which the ebay listing was created|
|nrOfPictures|The number of pictures in the ad|
|postalCode|The postal code for the location of the vehicle|
|lastSeenOnline|When the crawler saw this ad last online|

In [1]:
%%html
<style>
    table {float:left}
</style>

## Overview of Data

In [2]:
# Import Pandas and NumPy libraries
import pandas as pd
import numpy as np

In [3]:
# Import CSV file
autos = pd.read_csv('autos.csv', encoding='Latin-1')
autos.info()
autos.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   dateCrawled          50000 non-null  object
 1   name                 50000 non-null  object
 2   seller               50000 non-null  object
 3   offerType            50000 non-null  object
 4   price                50000 non-null  object
 5   abtest               50000 non-null  object
 6   vehicleType          44905 non-null  object
 7   yearOfRegistration   50000 non-null  int64 
 8   gearbox              47320 non-null  object
 9   powerPS              50000 non-null  int64 
 10  model                47242 non-null  object
 11  odometer             50000 non-null  object
 12  monthOfRegistration  50000 non-null  int64 
 13  fuelType             45518 non-null  object
 14  brand                50000 non-null  object
 15  notRepairedDamage    40171 non-null  object
 16  date

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

### Observations
- The dataset contains 20 columns, most of which are strings
- Column names are in `camelcase` instead of the standard `snakecase`
- There are 5 columns that **do not** contains 50000 possible non-null entries:
    - `vehicleType` contains 44905 non-null entries
    - `gearbox` contains 47320 non-null entries
    - `model` contains 47242 non-null entries
    - `fuelType` contains 45518 non-null entries
    - `notRepairedDamage` contains 40171 non-null entries

### Possible Changes that need to be made:
- `Price` includes the '$' which can be removed by changing the datatype from **object** to **int**
- `Odometer` has an **object** datatype and can be changed to **int**. Column needs to be renamed for better clarity


## Clean Columns
The labels for the columns need to change from `camelcase` to `snakecase` to conform to Python's style guide. In addition, Snakecase increases readability. 
Column names for the year and month of registration, not repaired damage, and date created need to be changed to make the column names more concise and specific.

In [4]:
# Print column names to copy and paste
print(autos.columns)

# Change from camelcase to snakecase
# Assign modified column names back to autos.column attribute
autos.columns = ['date_crawled', 'name', 'seller', 'offer_type', 'price', 'abtest',
       'vehicle_type', 'registration_year', 'gearbox', 'power_ps', 'model',
       'odometer', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'num_photos', 'postal_code',
       'last_seen']

autos.head()

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')


date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price   abtest vehicle_type  registration_year  \
0  privat    Angebot  $5,000  control          bus               2004   
1  privat    Angebot  $8,500  control    limousine               1997   
2  privat    Angebot  $8,990     test    limousine               2009   
3  privat    Angebot  $4,350  control   kleinwagen               2007   
4  privat    Angebot  $1,350     test        kombi               2003   

     gearbox  power_ps   model   odometer  registration_month fuel_type  \
0    manuell       158  andere  150,000km                   3       lpg   
1  automatik       286     7er  150,000km                   6    benzin   
2    manuell       102    golf   70,000km                   7    benzin   
3  automatik        71  fortwo   70,000km                   6    benzin   
4    manuell         0   focus  150,000km                   7    benzin   

        brand unrepaired_damage           ad_created  num_photos  postal_code  \
0     peugeot              nein  2016-03-26 00:00:00           0        79588   
1         bmw              nein  2016-04-04 00:00:00           0        71034   
2  volkswagen              nein  2016-03-26 00:00:00           0        35394   
3       smart              nein  2016-03-12 00:00:00           0        33729   
4        ford              nein  2016-04-01 00:00:00           0        39218   

             last_seen  
0  2016-04-06 06:45:54  
1  2016-04-06 14:45:08  
2  2016-04-06 20:15:37  
3  2016-03-15 03:16:28  
4  2016-04-01 14:38:50

## Initial Data Exploration and Cleaning

In [5]:
# Explore what additional data need to be cleaned and converted
autos.describe(include='all')

date_crawled         name  seller offer_type  price abtest  \
count                 50000        50000   50000      50000  50000  50000   
unique                48213        38754       2          2   2357      2   
top     2016-03-30 17:37:35  Ford_Fiesta  privat    Angebot     $0   test   
freq                      3           78   49999      49999   1421  25756   
mean                    NaN          NaN     NaN        NaN    NaN    NaN   
std                     NaN          NaN     NaN        NaN    NaN    NaN   
min                     NaN          NaN     NaN        NaN    NaN    NaN   
25%                     NaN          NaN     NaN        NaN    NaN    NaN   
50%                     NaN          NaN     NaN        NaN    NaN    NaN   
75%                     NaN          NaN     NaN        NaN    NaN    NaN   
max                     NaN          NaN     NaN        NaN    NaN    NaN   

       vehicle_type  registration_year  gearbox      power_ps  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fuel_type       brand unrepaired_damage  \
count       50000        50000.000000     45518       50000             40171   
unique         13                 NaN         7          40                 2   
top     150,000km                 NaN    benzin  volkswagen              nein   
freq        32424                 NaN     30107       10687             35232   
mean          NaN            5.723360       NaN         NaN               NaN   
std           NaN            3.711984       NaN         NaN               NaN   
min           NaN            0.000000       NaN         NaN               NaN   
25%           NaN            3.000000       NaN         NaN               NaN   
50%           NaN            6.000000       NaN         NaN               NaN   
75%           NaN            9.000000       NaN         NaN               NaN   
max           NaN           12.000000       NaN         NaN               NaN   

                 ad_created  num_photos   postal_code            last_seen  
count                 50000     50000.0  50000.000000                50000  
unique                   76         NaN           NaN                39481  
top     2016-04-03 00:00:00         NaN           NaN  2016-04-07 06:17:27  
freq                   1946         NaN           NaN                    8  
mean                    NaN         0.0  50813.627300                  NaN  
std                     NaN         0.0  25779.747957                  NaN  
min                     NaN         0.0   1067.000000                  NaN  
25%                     NaN         0.0  30451.000000                  NaN  
50%                     NaN         0.0  49577.000000                  NaN  
75%                     NaN         0.0  71540.000000                  NaN  
max                     NaN         0.0  99998.000000                  NaN

### Observations 
- `seller` and `offer_type` appear to be okay to be dropped since majority of the text column is uniform and thus does not provide any useful information.This is shown by both `seller` and `offer_type` having 2 unique values but the second only appearing 1 time.
    - `seller` - This would mean that 1 seller is responsible for 49999 while only 1 seller is responsible for the 1. 
    - `offer_type` - There are only 2 possible types of offers and with majority being 49999, the 1 other type would not give any significant information.
- the `num_photos` column is an additional column that we may need to investigate further as it contains all 0 values. We need to determined if the column will be useful.

In [6]:
autos['num_photos'].value_counts()

0    50000
Name: num_photos, dtype: int64

In [7]:
autos['seller'].value_counts()

privat        49999
gewerblich        1
Name: seller, dtype: int64

In [8]:
autos['offer_type'].value_counts()

Angebot    49999
Gesuch         1
Name: offer_type, dtype: int64

### Observations
- `num_photos` - Since we have confirmed that the column has all 0's, we can remove this column
- `seller` - We can see that 49999 of the 50000 are from private sellers. The 1 is from commercial. As a result, this information does not provide important information.
- `offer_type` - 49999 of the 50000 is a normal offer type while the 1 lone type is a formal request. These do not provide necessary information to be analyzed.

**Conclusion:** We can drop all 3 columns

## Cleaning - Dropping Unnecessary Columns

In [9]:
autos = autos.drop(['num_photos', 'seller', 'offer_type'], axis=1)

In [10]:
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

    price   abtest vehicle_type  registration_year    gearbox  power_ps  \
0  $5,000  control          bus               2004    manuell       158   
1  $8,500  control    limousine               1997  automatik       286   
2  $8,990     test    limousine               2009    manuell       102   
3  $4,350  control   kleinwagen               2007  automatik        71   
4  $1,350     test        kombi               2003    manuell         0   

    model   odometer  registration_month fuel_type       brand  \
0  andere  150,000km                   3       lpg     peugeot   
1     7er  150,000km                   6    benzin         bmw   
2    golf   70,000km                   7    benzin  volkswagen   
3  fortwo   70,000km                   6    benzin       smart   
4   focus  150,000km                   7    benzin        ford   

  unrepaired_damage           ad_created  postal_code            last_seen  
0              nein  2016-03-26 00:00:00        79588  2016-04-06 06:45:54  
1              nein  2016-04-04 00:00:00        71034  2016-04-06 14:45:08  
2              nein  2016-03-26 00:00:00        35394  2016-04-06 20:15:37  
3              nein  2016-03-12 00:00:00        33729  2016-03-15 03:16:28  
4              nein  2016-04-01 00:00:00        39218  2016-04-01 14:38:50

## Cleaning - Unwanted Symbols and Punctuation

Earlier, we mentioned that `price` contained a '$' symbol that we would need to remove. We also see that integers do not contain a comma so we would also remove the ','.  Lastly, we need to change the type to an int from an object. As such, we will proceed with these changes.

In [11]:
autos['price'] = (autos['price']
                          .str.replace('$','',regex=False)
                          .str.replace(',','',regex=False)
                          .astype({'price': 'int64'})
                         )
        
autos['price'].head()

0    5000
1    8500
2    8990
3    4350
4    1350
Name: price, dtype: int64

We will now proceed to the `odometer` column to remove the ',' and the 'km'

In [12]:
autos['odometer'] = (autos['odometer']
                          .str.replace(',','',regex=False)
                          .str.replace('km','',regex=False)
                          .astype({'odometer': 'int64'})
                           )
        
autos['odometer'].head()

0    150000
1    150000
2     70000
3     70000
4    150000
Name: odometer, dtype: int64

Now that we don't have the km label, we should be more specific with the column name. We will change the `odometer` column name to `odometer_km` to be more specific with our labeling.

In [13]:
autos = autos.rename(columns={'odometer' : 'odometer_km'})
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   price   abtest vehicle_type  registration_year    gearbox  power_ps  \
0   5000  control          bus               2004    manuell       158   
1   8500  control    limousine               1997  automatik       286   
2   8990     test    limousine               2009    manuell       102   
3   4350  control   kleinwagen               2007  automatik        71   
4   1350     test        kombi               2003    manuell         0   

    model  odometer_km  registration_month fuel_type       brand  \
0  andere       150000                   3       lpg     peugeot   
1     7er       150000                   6    benzin         bmw   
2    golf        70000                   7    benzin  volkswagen   
3  fortwo        70000                   6    benzin       smart   
4   focus       150000                   7    benzin        ford   

  unrepaired_damage           ad_created  postal_code            last_seen  
0              nein  2016-03-26 00:00:00        79588  2016-04-06 06:45:54  
1              nein  2016-04-04 00:00:00        71034  2016-04-06 14:45:08  
2              nein  2016-03-26 00:00:00        35394  2016-04-06 20:15:37  
3              nein  2016-03-12 00:00:00        33729  2016-03-15 03:16:28  
4              nein  2016-04-01 00:00:00        39218  2016-04-01 14:38:50

## Odometer Analysis

In [14]:
# Print minimum and maximum values
min_value = autos['odometer_km'].min()
max_value = autos['odometer_km'].max()
print("Minimum: {}\tMaximum: {}\n".format(min_value, max_value))

# Check unique values
print(autos['odometer_km'].unique().shape)

# Count number of values
autos['odometer_km'].value_counts()

Minimum: 5000	Maximum: 150000

(13,)


150000    32424
125000     5170
100000     2169
90000      1757
80000      1436
70000      1230
60000      1164
50000      1027
5000        967
40000       819
30000       789
20000       784
10000       264
Name: odometer_km, dtype: int64

### Observations
Here we can see that there are not any outliers shown. With just **13 unique odometer kilometers**, this insinuates that sellers are given rounded numbers to select from when it comes to reporting their odometer. The **issue** with this is the fact that there is a **cap at 150000 km**. This would mean that any car that has more than that (e.g. 300000 kg) can be listed at 150000 kg even though it has double that on its odometer. Because there is a significantly larger number of cars with 150000 km (65%) compared to the others being sold, we must be particularly cautious of this figure.


## Price Analysis

In [15]:
# Print minimum and maximum values
min_value = autos['price'].min()
max_value = autos['price'].max()
print("Minimum: {}\tMaximum: {}\n".format(min_value, max_value))

Minimum: 0	Maximum: 99999999



There seems to be a **notable outlier** in the `price` where the max price is nearly 100 million. This would be impossible so we can remove these listings. However, before doing so, we need to check if the max is the only outlier or if there are more unreasonably priced listings.

In [16]:
# Print statistics
print("\nStats:")
print(autos["price"].describe())

# Print unique values
print("\nUnique Values:")
print(autos["price"].unique().shape)

# Order by lowest Price
print("\nLowest priced listings")
print(autos["price"].value_counts().sort_index(ascending=True).head(20))

# Order by Highest Price
print("\nHighest priced listing")
autos["price"].value_counts().sort_index(ascending=False).head(20)


Stats:
count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price, dtype: float64

Unique Values:
(2357,)

Lowest priced listings
0     1421
1      156
2        3
3        1
5        2
8        1
9        1
10       7
11       2
12       3
13       2
14       1
15       2
17       3
18       1
20       4
25       5
29       1
30       7
35       1
Name: price, dtype: int64

Highest priced listing


99999999    1
27322222    1
12345678    3
11111111    2
10000000    1
3890000     1
1300000     1
1234566     1
999999      2
999990      1
350000      1
345000      1
299000      1
295000      1
265000      1
259000      1
250000      1
220000      1
198000      1
197000      1
Name: price, dtype: int64

### Observations:
For the **maximum**, we can see that the 99999999 listing is not the only unreasonable outlier. In order to keep the most reasonable listings, I will set the upper limit to 350000 as there is a significant gap between that and the next most expensive listing of 999990 (nearly $1 million).

For the **minimum**, there are 1421 listed at $0. However, given that eBay is an auction site, it is possible that these sellers want to just start as low as possible and allow bidders to bid up the price. As such, it would not be a bad idea to leave the 0's in.

Once we remove the outliers, let's see what we have.

In [17]:
autos = autos[autos['price'].between(0, 350001)]
autos['price'].describe()

count     49986.000000
mean       5721.525167
std        8983.617820
min           0.000000
25%        1100.000000
50%        2950.000000
75%        7200.000000
max      350000.000000
Name: price, dtype: float64

## Exploring Time

There are 5 columns with date values:

- `date_crawled` - Object
- `registration_month` - Int
- `registration_year` - Int
- `ad_created` - Object
- `last_seen` - Object

These are a combination of dates that were crawled, and dates with meta-information from the crawler. 

We'll explore each of these columns to learn more about the listings.

In [18]:
autos[['date_crawled', 'ad_created', 'last_seen']].head()

date_crawled           ad_created            last_seen
0  2016-03-26 17:47:46  2016-03-26 00:00:00  2016-04-06 06:45:54
1  2016-04-04 13:38:56  2016-04-04 00:00:00  2016-04-06 14:45:08
2  2016-03-26 18:57:24  2016-03-26 00:00:00  2016-04-06 20:15:37
3  2016-03-12 16:58:10  2016-03-12 00:00:00  2016-03-15 03:16:28
4  2016-04-01 14:38:50  2016-04-01 00:00:00  2016-04-01 14:38:50

In [19]:
print(autos['date_crawled']
         .str[:10]
         .value_counts(normalize=True, dropna=False)
         .sort_index(ascending=True)
        )

2016-03-05    0.025387
2016-03-06    0.013944
2016-03-07    0.035970
2016-03-08    0.033269
2016-03-09    0.033209
2016-03-10    0.032129
2016-03-11    0.032489
2016-03-12    0.036770
2016-03-13    0.015564
2016-03-14    0.036630
2016-03-15    0.033990
2016-03-16    0.029508
2016-03-17    0.031509
2016-03-18    0.013064
2016-03-19    0.034910
2016-03-20    0.037831
2016-03-21    0.037490
2016-03-22    0.032909
2016-03-23    0.032389
2016-03-24    0.029108
2016-03-25    0.031749
2016-03-26    0.032489
2016-03-27    0.031049
2016-03-28    0.034850
2016-03-29    0.034150
2016-03-30    0.033629
2016-03-31    0.031909
2016-04-01    0.033809
2016-04-02    0.035410
2016-04-03    0.038691
2016-04-04    0.036490
2016-04-05    0.013104
2016-04-06    0.003181
2016-04-07    0.001420
Name: date_crawled, dtype: float64


The site appears to have been crawled daily from early March to early April 2016. The amount of listings crawled per day does not have large swings.

In [20]:
print(autos['ad_created']
         .str[:10]
         .value_counts(normalize=True, dropna=False)
         .sort_index(ascending=True)
        )

2015-06-11    0.000020
2015-08-10    0.000020
2015-09-09    0.000020
2015-11-10    0.000020
2015-12-05    0.000020
                ...   
2016-04-03    0.038931
2016-04-04    0.036850
2016-04-05    0.011843
2016-04-06    0.003261
2016-04-07    0.001280
Name: ad_created, Length: 76, dtype: float64


There appears to be a significant increase in ads created in early April but the amount drops significantly after a couple of days.

In [21]:
print(autos['last_seen']
         .str[:10]
         .value_counts(normalize=True, dropna=False)
         .sort_index(ascending=True)
        )

2016-03-05    0.001080
2016-03-06    0.004421
2016-03-07    0.005362
2016-03-08    0.007582
2016-03-09    0.009843
2016-03-10    0.010763
2016-03-11    0.012524
2016-03-12    0.023807
2016-03-13    0.008983
2016-03-14    0.012804
2016-03-15    0.015884
2016-03-16    0.016445
2016-03-17    0.027928
2016-03-18    0.007422
2016-03-19    0.015744
2016-03-20    0.020706
2016-03-21    0.020726
2016-03-22    0.021586
2016-03-23    0.018585
2016-03-24    0.019565
2016-03-25    0.019205
2016-03-26    0.016965
2016-03-27    0.016024
2016-03-28    0.020846
2016-03-29    0.022326
2016-03-30    0.024847
2016-03-31    0.023827
2016-04-01    0.023106
2016-04-02    0.024887
2016-04-03    0.025367
2016-04-04    0.024627
2016-04-05    0.124275
2016-04-06    0.220982
2016-04-07    0.130957
Name: last_seen, dtype: float64


There is a relatively large increase for the last 3 days.

### Observations:
There does not appear to be any issues with the dates in this dataset

## Registration Year Analysis

We will now need to examine the `registration year`. We will need to ensure that the listings use valid years when filling in their `registration year`. We will first examine the distribution of these years.

In [22]:
autos['registration_year'].describe()

count    49986.000000
mean      2005.075721
std        105.727161
min       1000.000000
25%       1999.000000
50%       2003.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64


We can see here that the earliest registration in a listing is the year 1000. As one of the [first cars accessible to the massses was in 1908](https://en.wikipedia.org/wiki/Car),we can safely remove any vehicle `registration year` dated before this. 

Since these listings were from 2016, any year above 2016 would not be possible and as a result can be safely removed.

In [23]:
autos = autos[autos["registration_year"].between(1908,2016)]

In [24]:
print(autos['registration_year'].describe())
autos['registration_year'].value_counts(normalize=True)

count    48016.000000
mean      2002.806002
std          7.306212
min       1910.000000
25%       1999.000000
50%       2003.000000
75%       2008.000000
max       2016.000000
Name: registration_year, dtype: float64


2000    0.069852
2005    0.062792
1999    0.062438
2004    0.057002
2003    0.056794
          ...   
1948    0.000021
1931    0.000021
1938    0.000021
1939    0.000021
1952    0.000021
Name: registration_year, Length: 78, dtype: float64

### Observation:
From this new, up-to-date distribution, we can see that a vast majority of cars were registered within the past 20 years (from 2016). 

## Brand Analysis

In [25]:
autos["brand"].value_counts(normalize=True)

volkswagen        0.212117
bmw               0.110026
opel              0.108172
mercedes_benz     0.095364
audi              0.086409
ford              0.069768
renault           0.047359
peugeot           0.029532
fiat              0.025866
seat              0.018181
skoda             0.016036
mazda             0.015141
nissan            0.015099
citroen           0.013912
smart             0.013912
toyota            0.012475
sonstige_autos    0.010892
hyundai           0.009851
volvo             0.009247
mini              0.008643
mitsubishi        0.008143
honda             0.007852
kia               0.007102
alfa_romeo        0.006623
porsche           0.006102
suzuki            0.005915
chevrolet         0.005706
chrysler          0.003665
dacia             0.002562
daihatsu          0.002562
jeep              0.002249
subaru            0.002187
land_rover        0.002041
saab              0.001604
jaguar            0.001583
trabant           0.001562
daewoo            0.001500
r

### Observations:
- German brands represent 4 out the top 5 brands, and make up over 60% of the listings. 
- **Volkswagen** had nearly double the amount of cars listed for sale as the next two brands.
- Only the top 6 brands represent more than 5% of total listings

Since only 6 brands have a significant impact on the listings, we will focus on these 6.


## Price by Brand

We will need to analyze the 6 `brands` to see what kind of price variance exists among the different brands.

In [26]:
brand_counts = autos["brand"].value_counts(normalize=True)
pop_brands = brand_counts[brand_counts > .05].index
print(pop_brands)

Index(['volkswagen', 'bmw', 'opel', 'mercedes_benz', 'audi', 'ford'], dtype='object')


In [27]:
brand_mean_prices = {}

for brand in pop_brands:
    # As for loop runs, get each brand separately to average before assigning to respective bf
    brand_only = autos[autos["brand"] == brand]
    # While looping, assign that specific brand's price column's mean to mean_price
    mean_price = brand_only["price"].mean()
    # Assign value as int to brand name
    brand_mean_prices[brand] = int(mean_price)   

In [28]:
# Aggregate columns for easier visualization
bmp_series = pd.Series(brand_mean_prices)
pd.DataFrame(bmp_series, columns=["mean_price"])

mean_price
volkswagen           5231
bmw                  8102
opel                 2876
mercedes_benz        8485
audi                 9093
ford                 3652

### Observation:
For these top 5 `brands`, there are three distinct price brackets:
- **BMW**, **Mercedes Benz**, and **Audi** are the most expensive
    - This is unsurprising as these brands are known for their luxury vehicles and the price reflects that.
- **Opel** and **Ford** are the cheapest.
    - These two brands are known for pricing their automobiles for the average person at an affordable price.
- **Volkswagen** has its own bracket as an "in-between".
    - While they are not known to be budget cars, they market themselves as [durable and high-quality engineering](https://www.teamgunthervw.com/blog/why-are-volkswagen-vehicles-cheaper-than-other-german-cars/). 

## Price by Mileage

Now that we can see the **price variance** among different `brands`. We will need to account for the the mileage having an effect on the price variance among these brands. It is [common knowledge that the higher the mileage on a car, the lower the price](https://www.carsdirect.com/used-car-prices/used-car-price-per-mile-understanding-whats-exceptional). As a result, we will need to check the mileage to see if lower mileage is what contributes to these cars being more expensive.

In [29]:
brand_mean_mileage = {}

for brand in pop_brands:
    # As for loop runs, get each brand separately to average before assigning to respective brand
    brand_only = autos[autos["brand"] == brand]
    # While looping, assign that specific brand's kilometer column's mean to mean_mileage
    mean_mileage = brand_only["odometer_km"].mean()
    # Assign value as int to brand name
    brand_mean_mileage[brand] = int(mean_mileage)


In [30]:
# Aggregate columns for easier visualization
mean_mileage = pd.Series(brand_mean_mileage)
mean_prices = pd.Series(brand_mean_prices)
brand_info = pd.DataFrame(mean_mileage,columns=['mean_mileage'])
brand_info["mean_price"] = mean_prices
brand_info

mean_mileage  mean_price
volkswagen           128724        5231
bmw                  132431        8102
opel                 129223        2876
mercedes_benz        130856        8485
audi                 129287        9093
ford                 124068        3652

### Conclusion
The **average mileage** across these 6 `brands` is very close with each brand having a variance of at most 5%. This would indicate that the mileage did not play a large factor in the price variance among the different `brands` — that the `brands` themselves are a huge factor in the pricing.